room.cpp

```diff
@@ -119,6 +119,8 @@ class Room::Private
         void setLastReadEvent(User* u, const QString& eventId);
         rev_iter_pair_t promoteReadMarker(User* u, rev_iter_t newMarker);
 
+        void toJson(QJsonObject &out);
```

Comment:
Why not just return `QJsonObject` instead? If you're concerned about excess copying, just make sure to not break conditions for [RVO](https://en.wikipedia.org/wiki/Return_value_optimization).
And this method should be `const`.